### PR TITLE
Adapters

### PR DESCRIPTION
**Pull request by cromwellian**
- [x] Grammar for Adapter Declarations and Application
- [ ] Resolving adapters in Manifest.ts
- [ ] Kotlin Adapter representation in Plan
- [ ] Code-generator (recipe2plan) changes for emitting Adapters
- [ ] Implementation in Kotlin Handles (JoinHandle)
- [ ] Inline Adapters 
- [ ] Expand AdapterScopeExpression to RefinementExpression (+Piotr extensions)
- [ ] TS adapted handle implementation

### CODE REVIEW

File: src/runtime/manifest-parser.pegjs [hunk 1]
```diff
@@ -1581,6 +1584,84 @@ SchemaPrimitiveType
...
+  = paramName:AdapterParamName whiteSpace? ':' whiteSpace? type:ParticleHandleConnectionType {
+     return toAstNode<AstNode.AdapterParam>({
+        kind: 'adapter-param',
+        name: paramName,
+        type
+     });
+  }
+
+AdapterBodyDefinition
+  = name:('*' / upperIdent) whiteSpace? '{' multiLineSpace fields:AdapterFields? ','? multiLineSpace '}' {
```
Comment (piotrswigon) [inline src/runtime/manifest-parser.pegjs:1626]: Nice, we need to add an option of an optional trailing comma to schemas as well!

File: src/runtime/manifest-parser.pegjs [hunk 2]
```diff
@@ -1581,6 +1584,84 @@ SchemaPrimitiveType
     });
   }
 
+Adapter "an adapter, (e.g. adapter Foo(param: Person { name: Text }) => Friend { nickName: param.name } )"
+  = 'adapter' whiteSpace adapterName:AdapterName whiteSpace? '(' multiLineSpace? params:AdapterParamsDeclaration multiLineSpace? ')' whiteSpace? '=>' multiLineSpace? body:AdapterBodyDefinition eolWhiteSpace
```
Comment (piotrswigon) [inline src/runtime/manifest-parser.pegjs:1588]: Should we be opinionated on whether `(` has to be adjacent to the name? e.g. `ToFriend(...` vs `ToFriend (...`

I would err on the side of no whitespace, wdyt?

File: src/runtime/tests/manifest-parser-test.ts [hunk 3]
```diff
@@ -64,6 +64,51 @@ describe('manifest parser', () => {
...
+        places: map #locations
+        pairs: join (people, places) apply PairPerson(people, places)`);
+  });
+  it('parses adapter declarations with inline schema', () => {
+    parse(`
+      adapter ToFriend ( 
+        person: Person { name: Text, age: Number },
+        company: ~a with { name: Text, address: Text },
+        data: (Person {name: Text, age: Number }, Company { name: Text, address: Text })
+      ) => Friend { nickName: person.name, work: company.name, newdata: data.first }
```
Comment (piotrswigon) [inline src/runtime/tests/manifest-parser-test.ts:96]: You could add a test for 0 or multiple type names as well. e.g.

```
adapter Identifier(person: Person {name: Text}) => {ident: name}
adapter ToProduct(thing: {name: Text}) => Product Thing {name: Text}
```

```
adapter Identifier(person: Person {name: Text}) => {ident: name}
```

File: src/runtime/manifest-parser.pegjs (hunk 2 above)
Comment (cromwellian) [reply src/runtime/manifest-parser.pegjs:1588]: SGTM, I'll drop the whitespace. Seeing as how well opinionated formatting works for Go, I'm ok with not allowing arbitrary whitespace everywhere.

File: src/runtime/tests/manifest-parser-test.ts [hunk 4]
```diff
@@ -64,6 +64,51 @@ describe('manifest parser', () => {
...
+        person: Person { name: Text, age: Number },
+        place: ~a with { name: Text, address: Text },
+        data: (Person {name: Text, age: Number }, Company { name: Text, address: Text })
+      ) => PairedPerson { name: person.name, address: place.address, place: place.name }     
+      recipe
+        people: map #folks
+        places: map #locations
+        pairs: join (people, places) apply PairPerson(people, places)`);
+  });
+  it('parses adapter declarations with inline schema', () => {
```
Comment (piotrswigon) [inline src/runtime/tests/manifest-parser-test.ts:90]: we could also have a test for longer addressing chain, e.g.
```
adapter Inline(person: Person {
  name: Text,
  address: &Address {address: Text}
}) => Person {
  name: person.name,
  address: person.address.address
}
```

**Comment by piotrswigon** [comment]
Looks awesome, just minor comments :)

**Pull request merged**